File: algo/heap.c
#include<stdlib.h>
#include<stdio.h>


#define PQ_SIZE 1024

typedef struct pqueue {
    int q[PQ_SIZE+1];
    int n;
} pqueue_t;

int pq_parent(int n)
{
    if(n == 1) return(-1);
    else return((int)n/2);
}

int pq_young_child(int n) 
{
    return (2*n);
}

void pq_swap(pqueue_t *q, int a, int b)
{
    int t = q->q[a];
    q->q[a] = q->q[b];
    q->q[b] = t;
}

void bubble_up(pqueue_t *q, int p)
{
    if(pq_parent(p) == -1) return;

    if(q->q[pq_parent(p)] > q->q[p]) {
        pq_swap(q, p, pq_parent(p));
        bubble_up(q, pq_parent(p));
    }
}

void pq_insert(pqueue_t *q, int x)
{
    if(q->n >= PQ_SIZE)
        fprintf(stderr, "Warning: priority queue overflow insert x=%d\n",x);
    else {
        q->n = (q->n) + 1;
        q->q[q->n] = x;
        bubble_up(q, q->n);
    }
}

void pq_init(pqueue_t *q)
{
    q->n = 0;
}

void make_heap(pqueue_t *q, int s[], int n)
{
    int i;
    pq_init(q);
    for(i = 0; i <n; ++i)
        pq_insert(q, s[i]);
}

void bubble_down(pqueue_t *q, int p)
{
    int c = pq_young_child(p);
    int min_index = p;
    int i;

    for(i = 0; i <= 1; ++i)
        if((c+i) <= q->n) {
            if(q->q[min_index] > q->q[c+i]) min_index = c+i;
        }

    if(min_index != p) {
        pq_swap(q, p, min_index);
        bubble_down(q, min_index);
    }
}

int extract_min(pqueue_t *q)
{
    int min = -1;
    if(q->n <= 0) fprintf(stderr, "Warning: empty priority queue.\n");
    else {
        min = q->q[1];
        q->q[1] = q->q[q->n];
        --q->n;
        bubble_down(q, 1);
    }
    return min;
}

void make_heap2(pqueue_t *q, int s[], int n)
{
    int i;

    q->n = n;
    for(i = 0; i < n; ++i) q->q[i+1] = s[i];
    for(i = q->n; i >= 1; --i) bubble_down(q, i);
}

void my_heapsort(int s[], int n)
{
    int i;
    pqueue_t q;

    make_heap(&q, s, n);
    for(i = 0; i < n; ++i)
        s[i] = extract_min(&q);
}


#ifdef TEST
int
main(int argn, char *argv[])
{
    int i, n;
    pqueue_t q;

    n = atoi(argv[1]);

    pq_init(&q);

    for(i = 0; i < n; ++i) {
        pq_insert(&q, random()%PQ_SIZE);
    }

    for(i = 0; i < n; ++i)
        printf("%d\n", extract_min(&q));
    return 0;

}
#endif

File: python/heap.py
class pqueue:
    q = []
    def __init__(self, s = []):
#      Naive approach
#        for c in s:
#            self.insert(c)
#
        self.q.extend(s)
        for i in xrange(len(self.q)-1,-1,-1):
            self.bubble_down(i)

    def parent(self, n):
        if len(self.q) == 0:
            return -1
        else:
            return n/2

    def young_child(self, n):
        return 2*n

    def bubble_up(self, p):
        if self.parent(p) == -1:
            return

        m = self.parent(p)

        if self.q[m] > self.q[p]:
            self.q[m], self.q[p] = self.q[p], self.q[m]
            self.bubble_up(m)

    def insert(self, a):
        self.q.append(a)
        self.bubble_up(len(self.q)-1)

    def __init__(self, s = []):
        for c in s:
            self.insert(c)

    def bubble_down(self, p):
        c = self.young_child(p)
        min_index = p

        for i in [0,1]:
            if c+i < len(self.q):
                if self.q[min_index] > self.q[c+i]:
                    min_index = c+i

        if min_index != p:
            self.q[p], self.q[min_index] = self.q[min_index], self.q[p]
            self.bubble_down(min_index)

    def extract_min(self):
        min = self.q[0]
        self.q[0] = self.q[-1]
        del self.q[-1]
        self.bubble_down(0)
        return min

def heapsort(s):
    q = pqueue(s)
    for i in xrange(len(s)):
        s[i] = q.extract_min()
